### ISSUE TITLE
False positive `reportOperatorIssue` with recursive type

### ISSUE THREAD

**Issue by kryptan** [microsoft/pyright#10115]
**Describe the bug**
I get an `Operator "+" not supported for types "Value" and "list[int]"` error which is a false positive. With regular function with the same signature there is no error.

**Code or Screenshots**
```python
type IntoValue = int | list[IntoValue]

class Value:
    def __add__(self, r: IntoValue):
        pass

def f(a: Value, b: IntoValue):
    pass

if __name__ == "__main__":
    Value() + [1]
    f(Value(), [1])
```

![Image](https://github.com/user-attachments/assets/340c7674-609e-4fcf-9e40-d380b3f9db5e)

**VS Code extension or command-line**
Pylance 2025.3.2

**Comment by erictraut** [comment]
Pyright is working as intended here, so I don't consider this a bug.

Binary operators like `+` in Python are relatively complex and can invoke one of several different magic methods. For example, `+` may invoke `__add__` on the LHS operand or `__radd__` on the RHS operand. Because of this complexity, the types of both the LHS and RHS need to be evaluated upfront without the benefit of [bidirectional type inference](https://microsoft.github.io/pyright/#/type-inference?id=bidirectional-type-inference-expected-types). This means that default type inference rules are used. In your example, the default type inference rules for the expression `[1]` evaluate to `list[int]`. This argument type isn't assignable to `list[IntoValue]` because the type parameter for `list` is invariant.

By contrast, when `[1]` is evaluated as a simple argument within a call expression (like `f(Value(), [1]`), bidirectional type inference can be used, and this allows pyright to evaluate `[1]` as `list[IntoValue]` rather than `list[int]`.

You can modify your code in one of the following ways to allow it to type check.

```python
    temp: list[IntoValue] = [1]
    Value() + temp
```

Or

```python
    Value() + list[IntoValue]([1])
```
